### PR TITLE
Header hash list optimization

### PR DESCRIPTION
**Pull request by roman-khimov**
### Problem

Serialized access to hashes list.

### Solution

Use simple locks for it. Luckily there is exactly one place where it changes, all the other ones are readers.

### CODE REVIEW

File: pkg/core/blockchain.go [hunk 1]
```diff
@@ -490,65 +477,52 @@ func (bc *Blockchain) addHeaders(verify bool, headers ...*block.Header) (err err
...
-				return
-			}
+	buf := io.NewBufBinWriter()
+	bc.headerHashesLock.Lock()
+	defer bc.headerHashesLock.Unlock()
+	oldlen := len(bc.headerHashes)
+	var lastHeader *block.Header
+	for _, h := range headers {
+		if int(h.Index-1) >= len(bc.headerHashes) {
+			return fmt.Errorf("height of received header %d is higher than the current header %d", h.Index, len(bc.headerHashes))
```
Comment (AnnaShaleva) [inline pkg/core/blockchain.go]: Perhaps now we can get rid of this named `err` return value?

File: pkg/core/blockchain.go [hunk 2]
```diff
@@ -988,31 +954,35 @@ func (bc *Blockchain) HasBlock(hash util.Uint256) bool {
 }
 
 // CurrentBlockHash returns the highest processed block hash.
-func (bc *Blockchain) CurrentBlockHash() (hash util.Uint256) {
-	bc.headersOp <- func(headerList *HeaderHashList) {
-		hash = headerList.Get(int(bc.BlockHeight()))
+func (bc *Blockchain) CurrentBlockHash() util.Uint256 {
+	topBlock := bc.topBlock.Load()
+	if topBlock != nil {
+		if tb, ok := topBlock.(*block.Block); ok {
```
Comment (fyrchik) [inline pkg/core/blockchain.go:957]: Can it ever be `ok`?

**Review by AnnaShaleva — APPROVED** [review]

File: pkg/core/blockchain.go (hunk 2 above)
Comment (roman-khimov) [reply pkg/core/blockchain.go:957]: It can't, we only store valid Block pointers there.

**Review by fyrchik — APPROVED** [review]

File: pkg/core/blockchain.go (hunk 1 above)
Comment (roman-khimov) [reply pkg/core/blockchain.go]: Fixed.

**Pull request merged**